caas/kubernetes/provider/bootstrap_test.go

```diff
@@ -812,246 +788,140 @@ JUJU_DEV_FEATURE_FLAGS=developer-mode $JUJU_TOOLS_DIR/jujud machine --data-dir $
-
-		s.mockServiceAccounts.EXPECT().Get(gomock.Any(), controllerServiceAccount.Name, gomock.Any()).
-			Return(controllerServiceAccount, nil),
-		s.mockServiceAccounts.EXPECT().Update(gomock.Any(), controllerServiceAccount, gomock.Any()).
-			Return(controllerServiceAccount, nil),
-		s.mockClusterRoleBindings.EXPECT().Get(gomock.Any(), controllerServiceCRB.Name, gomock.Any()).
-			Return(nil, s.k8sNotFoundError()),
-		s.mockClusterRoleBindings.EXPECT().Patch(
-			gomock.Any(), controllerServiceCRB.Name, types.StrategicMergePatchType, gomock.Any(),
-			v1.PatchOptions{FieldManager: "juju"},
-		).Return(nil, s.k8sNotFoundError()),
-		s.mockClusterRoleBindings.EXPECT().Create(gomock.Any(), controllerServiceCRB, gomock.Any()).
-			Return(controllerServiceCRB, nil),
-
-		// ensure secret for caas-image-repo.
-		s.mockSecrets.EXPECT().Create(gomock.Any(), secretCAASImageRepo, gomock.Any()).Return(secretCAASImageRepo, nil),
-		s.mockServiceAccounts.EXPECT().Get(gomock.Any(), controllerServiceAccount.Name, gomock.Any()).
-			Return(controllerServiceAccount, nil),
-		s.mockServiceAccounts.EXPECT().Update(
-			gomock.Any(), controllerServiceAccountPatchedWithSecretCAASImageRepo, gomock.Any(),
-		).
-			Return(controllerServiceAccountPatchedWithSecretCAASImageRepo, nil),
-
-		// Check the operator storage exists.
-		// first check if <namespace>-<storage-class> exist or not.
-		s.mockStorageClass.EXPECT().Get(gomock.Any(), "controller-1-some-storage", v1.GetOptions{}).
-			Return(nil, s.k8sNotFoundError()),
-		// not found, fallback to <storage-class>.
-		s.mockStorageClass.EXPECT().Get(gomock.Any(), "some-storage", v1.GetOptions{}).
-			Return(&sc, nil),
-
-		s.mockStatefulSets.EXPECT().Create(gomock.Any(), statefulSetSpec, v1.CreateOptions{}).
-			DoAndReturn(func(_, _, _ interface{}) (*apps.StatefulSet, error) {
-				eventFirer()
-				return statefulSetSpec, nil
-			}),
-
-		s.mockEvents.EXPECT().List(
-			gomock.Any(),
-			listOptionsFieldSelectorMatcher("involvedObject.name=controller-0,involvedObject.kind=Pod"),
-		).Return(&core.EventList{}, nil),
-
-		s.mockEvents.EXPECT().List(
-			gomock.Any(),
-			listOptionsFieldSelectorMatcher("involvedObject.name=controller-0,involvedObject.kind=Pod"),
-		).DoAndReturn(func(...interface{}) (*core.EventList, error) {
-			podFirer()
-			return eventsPartial, nil
-		}),
-
-		s.mockEvents.EXPECT().List(
-			gomock.Any(),
-			listOptionsFieldSelectorMatcher("involvedObject.name=controller-0,involvedObject.kind=Pod"),
-		).Return(eventsDone, nil),
-
-		s.mockPods.EXPECT().Get(gomock.Any(), "controller-0", v1.GetOptions{}).
-			Return(podReady, nil),
-	)
-
 	errChan := make(chan error)
+	done := make(chan struct{})
+	s.AddCleanup(func(c *gc.C) {
+		close(done)
+	})
 	go func() {
 		errChan <- controllerStacker.Deploy()
 	}()
-
-	err = s.clock.WaitAdvance(3*time.Second, testing.LongWait, 1)
-	c.Assert(err, jc.ErrorIsNil)
+	go func() {
+		svcDone := time.After(0 * time.Second)
```

Comment:
Can we use a s.mockServices.Watch?

Reply:
Unfortunately, I don't think those event side effects are triggered automatically by the service gets created/updated... etc.
Just like the pods are not created even we created the Statefulset.

Reply:
Ah, my previous testing was wrong.
Changed to use the fake watcher now.